File: LeetCode/Java Solutions/Remove Nodes From Linked List/remove.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

 class Solution {
    public ListNode removeNodes(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        Stack<ListNode> stack = new Stack<>();
        ListNode current = head;
        while (current != null) {
            stack.push(current);
            current = current.next;
        }
        
        int maxVal = Integer.MIN_VALUE;
        ListNode newHead = null;
        while (!stack.isEmpty()) {
            ListNode node = stack.pop();
            if (node.val >= maxVal) {
                maxVal = node.val;
                node.next = newHead;
                newHead = node;
            }
        }
        
        return newHead;
    }
}

File: LeetCode/Python Solutions/Remove Nodes From Linked List/remove.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Edge case: if the list is empty or has one node
        if not head or not head.next:
            return head
        
        stack = []
        current = head
        # Push all nodes onto a stack
        while current:
            stack.append(current)
            current = current.next
            
        max_val = float('-inf')
        new_head = None
        # Traverse from the end (using the stack)
        while stack:
            node = stack.pop()
            if node.val >= max_val:
                max_val = node.val
                node.next = new_head
                new_head = node
        
        return new_head
